Unified approach to hosting dashboard redirects

markbiek opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->


## Proposed Changes

This proposes a more unified way of handling redirects in cases where a user enters the Stepper from either the `/ciab` or `/v2` dashboards and then needs to be redirected back. The goal is to come up with an easy way to handle those redirects that will work regardless of which dashboard the user was on. That way we don't have to worry about `/ciab` users being redirected to `/v2` or vice-versa.

We're adding a new `dashboard=` query param (currently should only be `ciab` or `v2`) in addition to the existing `redirect_to` param. Those two params together will be used to construct the full url that the user is redirected to.

The new `dashboard=` query param is also used to display the Woo logo when entering Stepper from the `/ciab` dashboard. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

For the `/v2` and `/ciab` dashboards, we need to check all paths that go into Stepper and then return to the correct location. This includes things like adding domains, upgrading to a plan, etc.

In most cases, the user will be redirected back to their starting location in the hosting dashboard. However some flows (like mapping a custom external domain) will take the user to the domain configuration settings instead.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

fushar reviewed (commented):
Apologies if I'm missing something, but wouldn't this be so much simpler if we just modify the existing `dashboardLink()` / `redirectToDashboardLink()` to return CIAB URL if we're on CIAB HD?

markbiek committed:
Add context-aware dashboard detection to link utilities

Introduces getCurrentDashboard() helper that detects which dashboard
(ciab or v2) the user came from by checking:
1. Query params (?dashboard=ciab or ?dashboard=v2)
2. Current path (starts with /ciab)
3. Referrer URL
4. Default to v2

Updates dashboardLink() to automatically prefix paths with /ciab
when the current context is the ciab dashboard, making redirect
URLs context-aware without requiring explicit param threading.

markbiek committed:
Pass dashboard param when entering Stepper flows from dashboard

Adds dashboard query param to all Stepper entry points (checkout,
plan-upgrade, domain flows, etc.) so the Stepper can determine
which dashboard the user came from and redirect back correctly.

Uses getCurrentDashboard() to detect the current dashboard context
and passes it via the dashboard param.

markbiek committed:
Show Woo branding in Stepper when coming from ciab dashboard

Adds useIsWooDashboard() hook that checks for ?dashboard=ciab
query param, and WooLogo SVG component matching the ciab dashboard
header.

When users enter Stepper flows from the ciab (WooCommerce) dashboard,
they now see consistent Woo branding instead of the WordPress logo.

markbiek committed:
Use context-aware dashboardLink() in Stepper flows

Updates Stepper flows to use dashboardLink() for redirect URLs:

- plan-upgrade: Uses dashboardLink('/sites') for fallback redirects
- woo-hosted-plans: Simplifies hardcoded /ciab/sites to /sites,
  relying on context-aware dashboardLink() to add prefix
- transferring-hosted-site-flow: Uses dashboardLink() for site redirect

This ensures users return to the correct dashboard after completing
Stepper flows.

markbiek committed:
Simplify dashboard detection to only explicitly detect ciab

- Remove SSR checks (these functions only run client-side)
- getDashboardFromPath(): Only detect ciab prefix, return null for
  all other paths (avoids fragile hardcoded path list)
- getDashboardFromReferrer(): Return null instead of v2 for non-ciab
  referrers (avoids misattributing external traffic)

Since v2 is the default, we only need to explicitly detect ciab.

markbiek committed:
Simplify useIsWooDashboard hook

Remove unnecessary useMemo and SSR check. URLSearchParams parsing
is cheap and this hook only runs client-side.

markbiek committed:
Handle edge cases in Stepper flow redirects

- transferring-hosted-site-flow: Handle undefined siteSlug by
  falling back to /sites list instead of /sites/undefined
- woo-hosted-plans: Update comment to reflect context-aware behavior

fushar reviewed (commented):
So the simpler way that I was suggesting is to not use a new `dashboard` query param. But, we simply construct the final `redirect_to` / `back_to` / whatever that is, with the correct absolute URL (to MSD / CIAB) via `redirectToDashboardLink()`. In that function (which calls `dashboardLink()`),  we can detect whether the "current dashboard" is CIAB by checking whether the current URL path starts with `/ciab`.

Would it make more sense?

fushar commented on packages/onboarding/src/step-container-v2/components/TopBar/TopBar.tsx:
```diff
@@ -4,6 +4,41 @@ import { isValidElement, type ReactElement, type ReactNode } from 'react';
+ * Check if we should show the Woo logo instead of WordPress logo.
+ * This is determined by the `dashboard=ciab` URL parameter.
+ */
+function useIsWooDashboard(): boolean {
```
This does not seem to be a hook.

markbiek replied:
Good point. I can clean that up.

zaguiini commented on packages/onboarding/src/step-container-v2/components/TopBar/TopBar.tsx:
```diff
@@ -4,6 +4,41 @@ import { isValidElement, type ReactElement, type ReactNode } from 'react';
+ * This is determined by the `dashboard=ciab` URL parameter.
+ */
+function useIsWooDashboard(): boolean {
+	const params = new URLSearchParams( window.location.search );
```
This won't work. The TopBar is rendered on the server as well, so we'll need to modify the [StepContainerV2Context](https://github.com/Automattic/wp-calypso/blob/trunk/packages/onboarding/src/step-container-v2/contexts/StepContainerV2Context.ts#L3) to include the Woo check.

Also, I'd prefer we pass the entire SVG within the context and keep the default logo as the default context value. We'll need to modify the context for it, but this change will massively improve maintainability and customization by removing conditional blocks!

Finally, we're using StepContainerV2 outside Stepper, such as on the login screen. We'll need to update `Layout` and `LayoutLoggedOut`, I think, so they also include the `StepContainerV2Context` because I don't think they do right now.

markbiek replied:
Ah, thanks for that context! I can get that fixed.

fushar commented on client/dashboard/utils/link.ts:
```diff
@@ -32,15 +32,96 @@ export function a4aLink( path: string ) {
+ * Priority: query param → current path → referrer → default (v2)
+ */
+export function getCurrentDashboard(): DashboardType {
+	return getDashboardFromQuery() ?? getDashboardFromPath() ?? getDashboardFromReferrer() ?? 'v2';
```
I think we should replace `v2` with `undefined`. This way, via `addQueryAags()`, we only add the `dashboard` param when we're in CIAB. Another reason is we're not using the `v2` terminology anymore in the "main MSD". What do you think?

markbiek replied:
Yeah, I think that makes sense. Since `v2` is the default, there's no reason to specify it by name.

markbiek committed:
Address feedback on logo changes

markbiek committed:
Fix redirect when adding a domain I own

markbiek committed:
Only add dashboard query param for CIAB

Change DashboardType from 'ciab' | 'v2' to 'ciab' | undefined so the
dashboard query param is only added to redirect URLs when on CIAB.
The main MSD dashboard is the default and doesn't need to be specified.

markbiek committed:
Fix transferring-hosted-site-flow test for dashboardLink usage

markbiek committed:
Fix TypeScript errors in getCurrentDashboard usage

zaguiini commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -120,13 +120,24 @@ export const FlowRenderer: React.FC< {
+	useEffect( () => {
+		const params = new URLSearchParams( window.location.search );
+		setIsWooDashboard( params.get( 'dashboard' ) === 'ciab' );
+	}, [] );
```
I think we can probably use a ref here. Or maybe nothing, just check for the parameter within the `useMemo` call, as the query parameters are supposed to be persisted between step changes.

zaguiini commented on client/layout/logged-out.jsx:
```diff
@@ -88,6 +89,24 @@ const LayoutLoggedOut = ( {
+		setIsWooDashboard( params.get( 'dashboard' ) === 'ciab' );
+	}, [] );
+
+	const stepContainerV2Context = useMemo(
```
Weird. Can we pass a partial context here? I thought you needed to provide everything. We can provide default values for flowName and stepname, then provide the recordTracksEvent function as there is a log in link at the top-right that gets tracked whenever the user clicks on it.

zaguiini commented on packages/onboarding/src/step-container-v2/components/logos/WooLogo.tsx:
```diff
@@ -0,0 +1,33 @@
+ */
+export function WooLogo() {
+	return (
+		<div className="step-container-v2__top-bar-wordpress-logo-wrapper">
```
I think we can remove this wrapper div as the logo is already being wrapped by this element.

zaguiini commented on packages/onboarding/src/step-container-v2/components/logos/index.ts:
```diff
@@ -0,0 +1 @@
+export { WooLogo } from './WooLogo';
```
Let's not add the logo to the onboarding package. We probably already have a logos one which is from where we're importing the wordmark and WordPress logo, so we can put this one there too.

zaguiini commented on packages/onboarding/src/step-container-v2/contexts/StepContainerV2Context.ts:
```diff
@@ -1,9 +1,10 @@
+	flowName?: string;
+	stepName?: string;
+	recordTracksEvent?: ( eventName: string, eventProperties: Record< string, unknown > ) => void;
+	logo?: ReactNode;
```
Oh I see, you made them optional. Yeah that works, but I prefer to have default values instead of making everything optional. Take a look at how we're doing it at the domain search package -- I think it's a better approach because we can then remove the optional chaining from within our code.

markbiek committed:
Address PR feedback on StepContainerV2 context and logo handling

- Update StepContainerV2Context to use default values instead of optional
  properties, removing the need for optional chaining throughout the code
- Move WooDashboardLogo to @automattic/components package where other
  logos live, and remove the wrapper div
- Simplify CIAB dashboard detection by checking query param directly in
  useMemo instead of useState/useEffect pattern
- Add recordTracksEvent to logged-out layout context for login link tracking
- Update Storybook decorator to match new context type requirements

zaguiini reviewed (approved):
I'm OK with this 👍🏻
